activesupport/lib/active_support/fork_tracker.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+module ActiveSupport
+  module ForkTracker # :nodoc:
+    module CoreExt
+      def fork(*)
+        if block_given?
+          super do
+            ForkTracker.check!
+            yield
+          end
+        else
+          unless pid = super
+            ForkTracker.check!
+          end
+          pid
+        end
+      end
+    end
+
+    @pid = Process.pid
+    @callbacks = []
+
+    class << self
+      def check!
+        if @pid != Process.pid
+          @callbacks.each(&:call)
+          @pid = Process.pid
+        end
+      end
+
+      def hook!
+        ::Object.prepend(CoreExt)
```

Comment:
This leaves `Kernel.fork` and `Kernel#fork` untouched... maybe we should cover those too. We can't help anyone who's already included Kernel, but it would at least catch anyone who includes it after we're hooked.

Reply:
> We can't help anyone who's already included Kernel

Actually we could if we monkey patched `Kernel` using an alias method chain. 

However I don't think `Kernel#fork` is a problem because `Kernel` is already included in Object. So the only way to bypass this patch would be to include `Kernel` in a `BasicObject` subclass, which really is a stretch IMO.

Good point about `Kernel.fork` though, I'll fix it.

Reply:
Including Kernel in a BasicObject is exactly what SimpleDelegator from the stdlib does. But it goes on to remove all the private instance methods of Kernel (which includes `fork`) to ensure messages pass through to the delegated object, so it's not a problem in the end.
https://github.com/ruby/ruby/blob/01e75b6be9987bf9ce0793643470e47a32871c38/lib/delegate.rb#L39-L40